pkg/runtime/bulk_subscriber.go

```diff
@@ -259,80 +266,98 @@ func (a *DaprRuntime) createEnvelopeAndInvokeSubscriber(ctx context.Context, psm
 			}
 			bulkMsg := pubsub.BulkMessage{
 				Entries:  entries,
-				Topic:    msg.Topic,
-				Metadata: msg.Metadata,
+				Topic:    psm.topic,
+				Metadata: psm.metadata,
 			}
-			if dlqErr := a.sendBulkToDeadLetter(ctx, psName, &bulkMsg, route.deadLetterTopic, entryIdIndexMap, nil, bulkSubDiag); dlqErr == nil {
+			if dlqErr := a.sendBulkToDeadLetter(ctx, psm.topic, &bulkMsg, route.deadLetterTopic, &entryIdIndexMap, nil, bulkSubDiag); dlqErr == nil {
 				// dlq has been configured and message is successfully sent to dlq.
-				for _, item := range psm.entries {
-					ind := (*entryIdIndexMap)[item.EntryId]
+				for _, item := range rawMsgEntries {
+					ind := (entryIdIndexMap)[item.EntryId]
 					setBulkResponseEntry(bulkResponses, ind, item.EntryId, nil)
 				}
 				return nil
 			}
 		}
-		bulkSubDiag.statusWiseDiag[string(pubsub.Retry)] += int64(len(psm.entries))
+		bulkSubDiag.statusWiseDiag[string(pubsub.Retry)] += int64(len(rawMsgEntries))
 
-		for _, item := range psm.entries {
-			ind := (*entryIdIndexMap)[item.EntryId]
+		for _, item := range rawMsgEntries {
+			ind := (entryIdIndexMap)[item.EntryId]
 			setBulkResponseEntry(bulkResponses, ind, item.EntryId, marshalErr)
 		}
 		return marshalErr
 	}
 	psm.data = da
-	psm.path = path
-	_, err = policy(func(ctx context.Context) (any, error) {
-		var pErr error
-		switch a.runtimeConfig.ApplicationProtocol {
-		case HTTPProtocol:
-			pErr = a.publishBulkMessageHTTP(ctx, &psm, bulkResponses, *entryIdIndexMap, bulkSubDiag)
-		case GRPCProtocol:
-			pErr = a.publishBulkMessageGRPC(ctx, &psm, bulkResponses, *entryIdIndexMap, bulkSubDiag)
-		default:
-			pErr = backoff.Permanent(errors.New("invalid application protocol"))
-		}
-		return nil, pErr
-	})
-	return err
-}
 
-// publishBulkMessageHTTP publishes bulk message to a subscriber using HTTP and takes care of corresponding responses.
-func (a *DaprRuntime) publishBulkMessageHTTP(ctx context.Context, msg *pubsubBulkSubscribedMessage,
-	bulkResponses *[]pubsub.BulkSubscribeResponseEntry, entryIdIndexMap map[string]int, bulkSubDiag *bulkSubIngressDiagnostics, //nolint:stylecheck
-) error {
-	spans := make([]trace.Span, len(msg.entries))
+	spans := make([]trace.Span, len(rawMsgEntries))
 
-	req := invokev1.NewInvokeMethodRequest(msg.path)
+	req := invokev1.NewInvokeMethodRequest(psm.path)
 	req.WithHTTPExtension(nethttp.MethodPost, "")
-	req.WithRawData(msg.data, contenttype.CloudEventContentType)
-	req.WithCustomHTTPMetadata(msg.metadata)
+	req.WithRawData(psm.data, contenttype.CloudEventContentType)
+	req.WithCustomHTTPMetadata(psm.metadata)
 
 	n := 0
-	for _, cloudEvent := range msg.cloudEvents {
+	for _, pubsubMsg := range psm.pubSubMessages {
+		cloudEvent := pubsubMsg.cloudEvent
 		if cloudEvent[pubsub.TraceIDField] != nil {
 			traceID := cloudEvent[pubsub.TraceIDField].(string)
 			sc, _ := diag.SpanContextFromW3CString(traceID)
-			spanName := fmt.Sprintf("pubsub/%s", msg.topic)
+			spanName := fmt.Sprintf("pubsub/%s", psm.topic)
```

Comment:
Haha, I saw you refactoring other instances of `fmt.Sprintf` to `+` as well. Is it purely due to perf. reasons?

<img width="710" alt="image" src="https://user-images.githubusercontent.com/22132944/206618463-587eae36-a8d6-4beb-bbfd-60eef7b580cc.png">

[[source](https://forum.golangbridge.org/t/fmt-sprintf-vs-string-concatenation/23006)]

Reply:
Yes. It's consistent with benchmarks I ran too in another PR. When it's just about string concatenation in a hot path, the perf difference is significant. Not just the ns/op, but the fact that the second has 0 heap allocations, while the first one has a number (in my test, I got 3) - and that's more work for the allocator and the GC.

Reply:
Good stuff! I wonder if it's worth adding it to our linter because it can get tedious to manually review it each time.

Reply:
I just looked but couldn't see an option in our linters to catch these kinds of things. I don't think we should forbid `fmt.Sprintf` entirely because there are many situations where it's fine: in tests and in other non-hot paths, and/or when there are very complex concatenations (such as where the benefits in terms of code readability outweight the perf impact)

Reply:
Yes, if we ever decide on configuring our linter, it should only be confined to simple concatenations.

Reply:
Changed.